api/connector/simpleconnector_test.go

```diff
@@ -0,0 +1,104 @@
+// Copyright 2024 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package connector
+
+import (
+	"github.com/juju/names/v5"
+	gc "gopkg.in/check.v1"
+
+	"github.com/juju/juju/api"
+	"github.com/juju/juju/testing"
+)
+
+type simpleConnectorSuite struct {
+	testing.BaseSuite
+}
+
+var _ = gc.Suite(&simpleConnectorSuite{})
+
+func (s *simpleConnectorSuite) TestNewSimpleRespectsClientCredentials(c *gc.C) {
+	tests := []struct {
+		name                    string
+		opts                    SimpleConfig
+		expectedError           string
+		expectedAPIInfo         api.Info
+		expectedDefaultDialOpts func() api.DialOpts
+	}{
+		{
+			name: "with username/password",
+			opts: SimpleConfig{
+				ControllerAddresses: []string{"some.host:9999"},
+				ModelUUID:           "some-uuid",
+				Username:            "some-username",
+				Password:            "some-password",
+			},
+			expectedAPIInfo: api.Info{
+				Addrs:    []string{"some.host:9999"},
+				ModelTag: names.NewModelTag("some-uuid"),
+				Tag:      names.NewUserTag("some-username"),
+				Password: "some-password",
+			},
+		},
+		{
+			name: "with client credentials",
+			opts: SimpleConfig{
+				ControllerAddresses: []string{"some.host:9999"},
+				ModelUUID:           "some-uuid",
+				ClientID:            "some-client-id",
+				ClientSecret:        "some-client-secret",
+			},
+			expectedAPIInfo: api.Info{
+				Addrs:    []string{"some.host:9999"},
+				ModelTag: names.NewModelTag("some-uuid"),
+			},
+			expectedDefaultDialOpts: func() api.DialOpts {
+				expected := api.DefaultDialOpts()
+				expected.LoginProvider = api.NewClientCredentialsLoginProvider("some-client-id", "some-client-secret")
+				return expected
+			},
+		},
+		{
+			name: "with username/password and client credentials; username/password takes over",
```

Comment:
Will we ever have a legitimate case where both client id / secret and username / password are both set? Should be instead add a validation check and error if both are supplied? Same for other potential unsupported combinations.

Reply:
I don't think so. I can add a simple if-statement to ensure username/password and client-id/client-secret are mutually exclusive. Is that what you're asking for?

Reply:
Done.